### DISCUSSION TITLE
How could we standardise tooling?

### DISCUSSION THREAD

**Discussion by sndrs** [guardian/csnx#428] [Team Posts]
- new module creation is long-winded
- release & versioning is manual, which adds friction and fragility
- example attempt at https://github.com/guardian/libs

**Comment by sndrs** [comment]
from @oliverlloyd's [original post](https://github.com/orgs/guardian/teams/client-side-infra/discussions/3):

> Where do people stand on https://github.com/guardian/libs?
> 
> I realise just going ahead and doing it is a bit pre-emptive of this discussion but actually there are already several examples of repos being spun up to share code and so `libs` was built on top of those and put out there with the purpose of being critic'd. I added the [storage module](https://github.com/guardian/libs/pull/1) as an example for how the library could be used.
> 
> Previous shared code repos built along the same lines include:
> https://github.com/guardian/ab-rendering
> https://github.com/guardian/atoms-rendering
> https://github.com/guardian/discussion-rendering/
> 
> `libs` is the result of the things we learnt with these but is really far from perfect; but perhaps at a point where we can all build on it together?

**Comment by JamieB-gu** [comment]
How does this relate to the [`example-rendering`](https://github.com/guardian/example-rendering) template repo?

**Comment by sndrs** [comment]
i'm not sure – i used it for https://github.com/guardian/commercial-web and ended up removing a lot of stuff i didn't need :)

(as an aside, i think the libs repo is a slight misdirection – it strikes me as definitely a good idea, but not ideal place for defining general best practice)

i suspect it will always be the case that a template repo will never be sufficient for new projects, always too much or not enough stuff in it. it's also very hard to trickle changes to best practice out of it.

things i think we need from this are:

- capture best practice tooling so we set it up once and it's revisable
- shouldn't get in the way if you don't need some things
- be able to easily contribute to it as technologies change/arrive
- enable instances of it to update in place

if we accept these, i'm inclined toward the [react-scripts](https://github.com/facebook/create-react-app/tree/master/packages/react-scripts) model we walked about before. 

so something like:

```json
{
    "scripts": { 
        "lint": "gu-mods lint" 
    },
    "devDependencies": { 
        "@guardian/modules": "1.x" 
    }
}
```

what do you think? are there things missing or problematic in that list? do scripts capture it all?

**Comment by JamieB-gu** [comment]
> i'm inclined toward the react-scripts model we walked about before

This sounds interesting! I'm not too familiar with this approach. Maybe we could put it on the agenda for the next meeting, or perhaps write up an issue detailing how it would work? 🤔

> i used it for https://github.com/guardian/commercial-web and ended up removing a lot of stuff i didn't need :)

Sounds similar to what I came up against when using it for `liveblog-rendering`, as mentioned on the [other thread](https://github.com/orgs/guardian/teams/client-side-infra/discussions/7/comments/1) 😬

**Comment by gtrufitt** [comment]
I really like https://plopjs.com/

**Comment by JamieB-gu** [comment]
Interesting name 🤔
